File: src/qfn2l/visitors.py
# Author:  mikolas
# Created on:  Tue Mar 17 2026
# Copyright (C) 2026, Mikolas Janota
#
"""Visitor classes for traversing and transforming z3 expressions."""

import typing
from abc import abstractmethod
from collections import defaultdict

from prefix import QLev
from utils import (
    ZERO,
    FALSE,
    TRUE,
    GetLevel,
    eval_mul,
    eval_sum,
    is_ite,
    is_numeral,
    is_symbolic_const,
    is_zero,
    is_one,
    is_true,
    is_false,
    mk_add,
    mk_and,
    mk_mul,
    mk_not,
    mk_or,
)
from z3 import (
    BoolRef,
    ExprRef,
    FreshConst,
    is_add,
    is_and,
    is_app,
    is_const,
    is_eq,
    is_mul,
    is_not,
    is_or,
    is_sub,
    simplify,
    substitute,
)


class SimpleVisit:
    """Naive memoized abstract class for visitor.

    Does not take into account quantifiers
    """

    def __init__(self):
        self._memo = {}

    @abstractmethod
    def visit_node(self, _e: ExprRef) -> typing.Any:
        """
        Abstract method: Subclasses MUST implement this method
        to visit nodes.
        """
        raise NotImplementedError("Not implemented")

    def __call__(self, a) -> typing.Any:
        return self.visit(a)

    def recurse(self, a: ExprRef) -> ExprRef:
        """Identity function for transformer visitors."""
        if not is_app(a) or a.num_args() == 0:
            return a
        rc = []
        change = False
        for ix in range(a.num_args()):
            c: ExprRef = a.arg(ix)
            nc = self(c)
            if not c.eq(nc):
                change = True
            rc.append(nc)
        if not change:
            return a
        return a.decl()(rc)

    def visit(self, a) -> typing.Any:
        """A memoized wrapper around visit_node."""
        if a in self._memo:
            return self._memo[a]
        res = self.visit_node(a)
        self._memo[a] = res
        return res


class Contains(SimpleVisit):
    def __init__(self, cs: set):
        super().__init__()
        self.cs = cs

    def visit_node(self, a):
        if a in self.cs:
            return True
        if not is_app(a):
            return False
        for i in range(a.num_args()):
            if self(a.arg(i)):
                return True
        return False


class HasUninterpreted(SimpleVisit):
    """Check for a given term whether it contains an uninterpreted constant."""

    def __init__(self):
        SimpleVisit.__init__(self)

    def __call__(self, a) -> bool:
        return super().__call__(a)

    def visit_node(self, e: ExprRef) -> bool:
        return (
            is_symbolic_const(e)
            or is_app(e)
            and any(self(e.arg(child_ix)) for child_ix in range(e.num_args()))
        )


class SimplePropagate(SimpleVisit):
    def __init__(self):
        super().__init__()

    def __call__(self, a) -> ExprRef:
        return super().__call__(a)

    def propagate(self, pos: bool, a: ExprRef):
        """Propagate simple equalities within an And (pos=True) or Or
        (pos=False).

        For And: extracts `x == c` where x is symbolic, substitutes into siblings.
        For Or: extracts `not(x == c)` (i.e. x != c), substitutes into siblings.

        Only handles equalities between constants (symbolic or literal).
        """

        def get_eq(e: ExprRef):
            """Extract equality (lhs, rhs) from a literal, with symbolic const
            on left.

            For pos=True: looks for `x == c` For pos=False: looks for
            `not(x == c)` Returns None if not a suitable equality.
            """
            if not pos:
                # Under Or, we need negated equalities: not(x == c)
                if is_not(e):
                    e = e.arg(0)
                else:
                    return None

            if not is_eq(e) or e.num_args() != 2:
                return None
            lhs, rhs = e.arg(0), e.arg(1)
            # Only handle equalities between constants (no complex terms)
            if not is_const(lhs) or not is_const(rhs):
                return None
            # Ensure symbolic constant is on the left
            if not is_symbolic_const(lhs):
                lhs, rhs = rhs, lhs
            if not is_symbolic_const(lhs):
                # Neither side is symbolic (e.g., 5 == 3)
                return None
            return lhs, rhs

        assert not pos or is_and(a)
        assert pos or is_or(a)

        chs: list[BoolRef] = a.children()
        eqs = set()
        i = 0
        # Extract equality literals from children (remove them from chs)
        while i < len(chs):
            if (eq := get_eq(chs[i])) is not None:
                # Swap-remove: replace current with last element
                chs[i] = chs[-1]
                chs.pop()
                eqs.add(eq)
            else:
                i += 1

        if not eqs:
            return a

        # Build substitution list, applying earlier substitutions to later rhs values
        # This handles chains like: x == y, y == 5 -> x maps to 5, y maps to 5
        substitution = []
        for lhs, rhs in eqs:
            rhs = substitute(rhs, *substitution)
            substitution.append((lhs, rhs))

        # Apply substitution to all remaining children
        chs = [substitute(ch, *substitution) for ch in chs]

        # Rebuild equality literals (skip trivial ones like x == x after substitution)
        eq_lits: list[BoolRef] = [
            lhs == rhs if pos else mk_not(lhs == rhs)
            for lhs, rhs in substitution
            if not lhs.eq(rhs)
        ]
        new_ch: list[BoolRef] = chs + eq_lits
        new_ch.sort(key=lambda c: c.get_id())
        return mk_and(*new_ch) if pos else mk_or(*new_ch)

    def visit_node(self, a):
        rv = self.recurse(a)
        if is_and(rv):
            return self.propagate(True, rv)
        if is_or(rv):
            return self.propagate(False, rv)
        return rv


class SimpleSimplify(SimpleVisit):
    def __init__(self):
        super().__init__()

    def visit_ite(self, a: ExprRef):
        cs = a.children()
        assert len(cs) == 3
        c, t, e = cs
        if is_true(c):
            return t
        if is_false(c):
            return e
        return a

    def visit_sub(self, a: ExprRef):
        return a.arg(0) if a.num_args() == 2 and is_zero(a.arg(1)) else a

    def visit_add(self, a: ExprRef):
        chs = a.children()
        if len(chs) == 1:
            return chs[0]

        coeffs = []
        others = []
        while chs:
            c = chs.pop()
            if is_numeral(c):
                coeffs.append(c)
            elif is_add(c):
                chs.extend(c.children())
            else:
                others.append(c)
        c = eval_sum(*coeffs)
        others.sort(key=lambda c: c.get_id())
        return mk_add(*others) if is_zero(c) else mk_add(c, *others)

    def visit_or(self, a: ExprRef):
        chs = a.children()
        nchs = []
        while chs:
            c = chs.pop()
            if is_true(c):
                return TRUE
            if is_or(c):
                chs.extend(c.children())
            elif not is_false(c):
                nchs.append(c)
        nchs.sort(key=lambda c: c.get_id())
        return mk_or(*nchs)

    def visit_and(self, a: ExprRef):
        chs = a.children()
        nchs = []
        while chs:
            c = chs.pop()
            if is_false(c):
                return FALSE
            if is_and(c):
                chs.extend(c.children())
            elif not is_true(c):
                nchs.append(c)
        nchs.sort(key=lambda c: c.get_id())
        return mk_and(*nchs)

    def visit_mul(self, a: ExprRef):
        chs = a.children()
        if len(chs) == 1:
            return chs[0]

        coeffs = []
        others = []
        while chs:
            c = chs.pop()
            if is_one(c):
                continue
            if is_zero(c):
                return ZERO
            if is_numeral(c):
                coeffs.append(c)
            elif is_mul(c):
                chs.extend(c.children())
            else:
                others.append(c)
        c = eval_mul(*coeffs)
        if is_zero(c):
            return ZERO
        others.sort(key=lambda c: c.get_id())
        if is_one(c):
            return mk_mul(*others)
        return mk_mul(c, *others)

    def visit_node(self, a):
        rv = self.recurse(a)
        if is_mul(rv):
            return self.visit_mul(rv)
        if is_add(rv):
            return self.visit_add(rv)
        if is_ite(rv):
            return self.visit_ite(rv)
        if is_sub(rv):
            return self.visit_sub(rv)
        if is_and(rv):
            return self.visit_and(rv)
        if is_or(rv):
            return self.visit_or(rv)
        if is_app(rv):
            n = rv.num_args()
            if n > 0 and all(is_numeral(rv.arg(i)) for i in range(n)):
                return simplify(rv)
        return rv


class MakeDefs(SimpleVisit):
    """Make definitions for complex expressions."""

    def __init__(self):
        SimpleVisit.__init__(self)
        self.hu = HasUninterpreted()
        self._definitions: dict[ExprRef, ExprRef] = {}

    def make(self, in_prefix, formula) -> tuple[list[QLev], ExprRef]:
        new_formula = self(formula)
        prefix = in_prefix[:]
        const2lev = {v: lev for lev, qlev in enumerate(prefix) for v in qlev.vars()}
        levs = GetLevel(const2lev)
        for t, v in self._definitions.items():
            new_level = -1
            term_level = levs(t)
            for lev in range(term_level, len(prefix)):
                qlev = prefix[lev]
                if qlev.is_exists():
                    qlev.add_var(v)
                    new_level = lev
                    break
            if new_level < 0:
                new_level = len(prefix)
                prefix.append(QLev(is_forall=False, vs=[v]))
            const2lev[v] = new_level
        return prefix, mk_and(
            new_formula, *[k == v for k, v in self._definitions.items()]
        )

    def __call__(self, a) -> ExprRef:
        return super().__call__(a)

    def _mk_def(self, t):
        d = self._definitions.get(t)
        if d is not None:
            return d
        nc = FreshConst(t.sort())
        self._definitions[t] = nc
        return nc

    def visit_node(self, init_t: ExprRef):
        t = self.recurse(init_t)
        if not t.eq(init_t):
            return self(t)
        if not is_mul(t):
            return t
        children = t.children()
        usymbols: int = sum(1 for c in children if self.hu(c))
        if usymbols < 2:
            return t
        coeffs = []
        splits = defaultdict(list)
        for c in children:
            if not self.hu(c):
                coeffs.append(c)
            else:
                x = c if is_const(c) else self._mk_def(c)
                splits[x].append(x)
        splits = list(splits.values())
        coeff = eval_mul(*coeffs)
        if is_zero(coeff):
            return ZERO
        while len(splits) > 2:
            a = splits.pop(0)
            b = splits.pop(0)
            d = self._mk_def(mk_mul(*a, *b))
            splits.append([d])
        assert 0 < len(splits) < 3
        return (
            mk_mul(coeff, *splits[0])
            if len(splits) == 1
            else mk_mul(coeff, *splits[0], *splits[1])
        )

File: src/qfn2l_cppz3/visitors.cpp
#include "visitors.h"
#include <algorithm>
#include <unordered_map>

// ── TermTransformer ───────────────────────────────────────────────────────────
// Iterative post-order traversal so visit_node() sees fully-transformed children.
Term TermTransformer::operator()(const Term& root) {
    {
        auto it = _memo.find(root);
        if (it != _memo.end()) return it->second;
    }
    std::vector<std::pair<Term, bool>> stk;
    stk.push_back({root, false});
    while (!stk.empty()) {
        Term t     = stk.back().first;
        bool pushed = stk.back().second;
        if (_memo.count(t)) {
            stk.pop_back();
            continue;
        }
        if (!pushed) {
            stk.back().second = true;
            if (t.is_app()) {
                for (unsigned i = 0; i < t.num_args(); ++i)
                    if (!_memo.count(t.arg(i)))
                        stk.push_back({t.arg(i), false});
            }
        } else {
            stk.pop_back();
            _memo.insert_or_assign(t, visit_node(t));
        }
    }
    return _memo.at(root);
}

Term TermTransformer::recurse(const Term& t) {
    if (!t.is_app() || t.num_args() == 0) return t;
    TermVec new_args;
    bool changed = false;
    for (unsigned i = 0; i < t.num_args(); ++i) {
        Term nc = (*this)(t.arg(i));
        if (nc.id() != t.arg(i).id()) changed = true;
        new_args.push_back(nc);
    }
    if (!changed) return t;
    return rebuild(_ctx, t, new_args);
}

// ── FlattenMul ────────────────────────────────────────────────────────────────
Term FlattenMul::visit_node(const Term& t) {
    Term t2 = recurse(t);
    if (!is_mul(t2)) return t2;
    TermVec flat;
    bool has_mul_child = false;
    for (unsigned i = 0; i < t2.num_args(); ++i) {
        Term c = t2.arg(i);
        if (is_mul(c)) {
            has_mul_child = true;
            for (unsigned j = 0; j < c.num_args(); ++j)
                flat.push_back(c.arg(j));
        } else {
            flat.push_back(c);
        }
    }
    return has_mul_child ? mk_mul(_ctx, flat) : t2;
}

// ── TermPredicate ─────────────────────────────────────────────────────────────
bool TermPredicate::operator()(const Term& root) {
    {
        auto it = _memo.find(root);
        if (it != _memo.end()) return it->second;
    }
    std::vector<std::pair<Term, bool>> stk;
    stk.push_back({root, false});
    while (!stk.empty()) {
        Term t     = stk.back().first;
        bool pushed = stk.back().second;
        if (_memo.count(t)) {
            stk.pop_back();
            continue;
        }
        if (!pushed) {
            stk.back().second = true;
            for (unsigned i = 0; i < t.num_args(); ++i)
                if (!_memo.count(t.arg(i)))
                    stk.push_back({t.arg(i), false});
        } else {
            stk.pop_back();
            _memo[t] = visit_node(t);
        }
    }
    return _memo.at(root);
}

// ── HasUninterpreted ──────────────────────────────────────────────────────────
bool HasUninterpreted::visit_node(const Term& t) {
    if (is_symbolic_const(t)) return true;
    for (unsigned i = 0; i < t.num_args(); ++i)
        if ((*this)(t.arg(i))) return true;
    return false;
}

// ── Contains ─────────────────────────────────────────────────────────────────
bool Contains::visit_node(const Term& t) {
    if (_cs.count(t)) return true;
    for (unsigned i = 0; i < t.num_args(); ++i)
        if ((*this)(t.arg(i))) return true;
    return false;
}

// ── SimpleSimplify ────────────────────────────────────────────────────────────
Term SimpleSimplify::visit_add(const Term& t) {
    std::vector<Term> stack = get_children(t);
    TermVec coeffs, others;
    while (!stack.empty()) {
        auto c = stack.back(); stack.pop_back();
        if (is_value(c))
            coeffs.push_back(c);
        else if (is_add(c))
            for (unsigned i = 0; i < c.num_args(); ++i) stack.push_back(c.arg(i));
        else
            others.push_back(c);
    }
    Term c = eval_sum(_ctx, coeffs);
    std::sort(others.begin(), others.end(),
              [](auto& a, auto& b) { return a.hash() < b.hash(); });
    if (is_zero(_ctx, c)) return mk_add(_ctx, others);
    TermVec all;
    all.push_back(c);
    all.insert(all.end(), others.begin(), others.end());
    return mk_add(_ctx, all);
}

Term SimpleSimplify::visit_mul(const Term& t) {
    std::vector<Term> stack = get_children(t);
    TermVec coeffs, others;
    while (!stack.empty()) {
        auto c = stack.back(); stack.pop_back();
        if (is_one(_ctx, c))       continue;
        if (is_zero(_ctx, c))      return _ctx.ZERO;
        if (is_value(c))           coeffs.push_back(c);
        else if (is_mul(c))
            for (unsigned i = 0; i < c.num_args(); ++i) stack.push_back(c.arg(i));
        else
            others.push_back(c);
    }
    Term c = eval_mul(_ctx, coeffs);
    if (is_zero(_ctx, c)) return _ctx.ZERO;
    std::sort(others.begin(), others.end(),
              [](auto& a, auto& b) { return a.hash() < b.hash(); });
    if (is_one(_ctx, c)) return mk_mul(_ctx, others);
    TermVec all;
    all.push_back(c);
    all.insert(all.end(), others.begin(), others.end());
    return mk_mul(_ctx, all);
}

Term SimpleSimplify::visit_or(const Term& t) {
    std::vector<Term> stack = get_children(t);
    TermVec nchs;
    while (!stack.empty()) {
        auto c = stack.back(); stack.pop_back();
        if (is_true(_ctx, c))  return _ctx.TRUE_T;
        if (is_or(c))
            for (unsigned i = 0; i < c.num_args(); ++i) stack.push_back(c.arg(i));
        else if (!is_false(_ctx, c))
            nchs.push_back(c);
    }
    std::sort(nchs.begin(), nchs.end(),
              [](auto& a, auto& b) { return a.hash() < b.hash(); });
    return mk_or(_ctx, nchs);
}

Term SimpleSimplify::visit_and(const Term& t) {
    std::vector<Term> stack = get_children(t);
    TermVec nchs;
    while (!stack.empty()) {
        auto c = stack.back(); stack.pop_back();
        if (is_false(_ctx, c)) return _ctx.FALSE_T;
        if (is_and(c))
            for (unsigned i = 0; i < c.num_args(); ++i) stack.push_back(c.arg(i));
        else if (!is_true(_ctx, c))
            nchs.push_back(c);
    }
    std::sort(nchs.begin(), nchs.end(),
              [](auto& a, auto& b) { return a.hash() < b.hash(); });
    return mk_and(_ctx, nchs);
}

Term SimpleSimplify::visit_sub(const Term& t) {
    auto chs = get_children(t);
    if (chs.size() == 2 && is_zero(_ctx, chs[1])) return chs[0];
    return t;
}

Term SimpleSimplify::visit_ite(const Term& t) {
    auto chs = get_children(t);
    assert(chs.size() == 3);
    if (is_true(_ctx, chs[0]))  return chs[1];
    if (is_false(_ctx, chs[0])) return chs[2];
    return t;
}

Term SimpleSimplify::visit_node(const Term& t) {
    Term rv = recurse(t);
    if (is_mul(rv)) return visit_mul(rv);
    if (is_add(rv)) return visit_add(rv);
    if (is_ite(rv)) return visit_ite(rv);
    if (is_sub(rv)) return visit_sub(rv);
    if (is_and(rv)) return visit_and(rv);
    if (is_or(rv))  return visit_or(rv);
    return rv;
}

// ── SimplePropagate ───────────────────────────────────────────────────────────
Term SimplePropagate::propagate(bool pos, const Term& node) {
    TermVec chs = get_children(node);
    using EqPair = std::pair<Term, Term>;
    std::vector<EqPair> eqs;

    for (int i = static_cast<int>(chs.size()) - 1; i >= 0; --i) {
        Term ch = chs[i];
        if (!pos) {
            if (!is_not(ch)) continue;
            ch = ch.arg(0);
        }
        if (!is_eq(ch)) continue;
        auto lhs = ch.arg(0), rhs = ch.arg(1);
        if (!is_symbolic_const(lhs) && !is_symbolic_const(rhs)) continue;
        if (!is_symbolic_const(lhs)) std::swap(lhs, rhs);
        if (!is_symbolic_const(lhs)) continue;
        eqs.push_back({lhs, rhs});
        chs[i] = chs.back();
        chs.pop_back();
    }
    if (eqs.empty()) return node;

    TermMap subst;
    for (auto& [lhs, rhs] : eqs) {
        Term new_rhs = do_substitute(_ctx, rhs, subst);
        subst.insert_or_assign(lhs, new_rhs);
    }
    for (auto& ch : chs)
        ch = do_substitute(_ctx, ch, subst);

    TermVec eq_lits;
    for (auto& [lhs, rhs] : subst) {
        if (z3::eq(lhs, rhs)) continue;
        Term eq = (lhs == rhs);
        eq_lits.push_back(pos ? eq : mk_not(_ctx, eq));
    }
    TermVec all_chs;
    all_chs.insert(all_chs.end(), chs.begin(), chs.end());
    all_chs.insert(all_chs.end(), eq_lits.begin(), eq_lits.end());
    std::sort(all_chs.begin(), all_chs.end(),
              [](auto& a, auto& b) { return a.hash() < b.hash(); });
    return pos ? mk_and(_ctx, all_chs) : mk_or(_ctx, all_chs);
}

Term SimplePropagate::visit_node(const Term& t) {
    Term rv = recurse(t);
    if (is_and(rv)) return propagate(true,  rv);
    if (is_or(rv))  return propagate(false, rv);
    return rv;
}

// ── MakeDefs ──────────────────────────────────────────────────────────────────
MakeDefs::MakeDefs(const Ctx& ctx) : TermTransformer(ctx), _hu(ctx) {}

Term MakeDefs::mk_def(const Term& t) {
    auto it = _definitions.find(t);
    if (it != _definitions.end()) return it->second;
    Term nc = _ctx.fresh_symbol(t.get_sort());
    _definitions.insert_or_assign(t, nc);
    return nc;
}

Term MakeDefs::visit_node(const Term& init_t) {
    Term t = recurse(init_t);
    if (!is_mul(t)) return t;

    TermVec children = get_children(t);
    int usymbols = 0;
    for (auto& c : children)
        if (_hu(c)) ++usymbols;
    if (usymbols < 2) return t;

    TermVec coeffs;
    std::unordered_map<Term, TermVec, ExprHash, ExprEq> splits;
    std::function<void(const Term&)> add = [&](const Term& c) {
        if (!_hu(c)) {
            coeffs.push_back(c);
        } else if (is_symbol(c)) {
            splits[c].push_back(c);
        } else if (is_mul(c)) {
            for (unsigned i = 0; i < c.num_args(); ++i) add(c.arg(i));
        } else {
            Term x = mk_def(c);
            splits[x].push_back(x);
        }
    };
    for (auto& c : children) add(c);

    Term coeff = eval_mul(_ctx, coeffs);
    if (is_zero(_ctx, coeff)) return _ctx.ZERO;

    std::vector<TermVec> split_list;
    for (auto& [_, v] : splits) split_list.push_back(v);

    while (split_list.size() > 2) {
        auto a = split_list[0]; split_list.erase(split_list.begin());
        auto b = split_list[0]; split_list.erase(split_list.begin());
        TermVec ab;
        ab.insert(ab.end(), a.begin(), a.end());
        ab.insert(ab.end(), b.begin(), b.end());
        Term d = mk_def(mk_mul(_ctx, ab));
        split_list.push_back({d});
    }

    assert(!split_list.empty() && split_list.size() <= 2);
    TermVec result_factors = {coeff};
    for (auto& grp : split_list)
        result_factors.insert(result_factors.end(), grp.begin(), grp.end());
    return mk_mul(_ctx, result_factors);
}

std::pair<Prefix, Term> MakeDefs::make(const Prefix& in_prefix, const Term& formula) {
    Term new_formula = (*this)(formula);
    Prefix prefix = in_prefix;
    for (auto& [t, v] : _definitions)
        prefix[0].add_var(v);

    TermVec body_parts = {new_formula};
    for (auto& [t, v] : _definitions)
        body_parts.push_back(t == v);
    return {prefix, mk_and(_ctx, body_parts)};
}
